File: profiler/UseVirtualDevice/UseVirtualDevice.cpp
/*******************************************************************************
 *BSD 3-Clause License
 *
 *Copyright (c) 2016-2024, Mech-Mind Robotics
 *All rights reserved.
 *
 *Redistribution and use in source and binary forms, with or without
 *modification, are permitted provided that the following conditions are met:
 *
 *1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *3. Neither the name of the copyright holder nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 *THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 *FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 *OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

/*
With this sample, you can acquire the profile data stored in a virtual device, generate the
intensity image and depth map, and save the images.
*/

#include <opencv2/imgcodecs.hpp>
#include <iostream>
#include <thread>
#include <chrono>
#include <mutex>
#include "profiler/VirtualProfiler.h"
#include "profiler/api_util.h"
#include "profiler/parameters/ScanParameters.h"

namespace {
std::mutex kMutex;

// Define the callback function for retrieving the profile data
void callbackFunc(const mmind::eye::ProfileBatch& batch, void* pUser)
{
    std::unique_lock<std::mutex> lock(kMutex);
    auto* outPutBatch = static_cast<mmind::eye::ProfileBatch*>(pUser);
    outPutBatch->append(batch);
}

void saveMap(const mmind::eye::ProfileBatch& batch, int lineCount, int width,
             const std::string& path)
{
    if (batch.isEmpty()) {
        std::cout
            << "The depth map cannot be saved because the batch does not contain any profile data."
            << std::endl;
        return;
    }
    cv::imwrite(path, cv::Mat(lineCount, width, CV_32FC1, batch.getDepthMap().data()));
}

void saveIntensity(const mmind::eye::ProfileBatch& batch, int lineCount, int width,
                   const std::string& path)
{
    if (batch.isEmpty()) {
        std::cout
            << "The intensity cannot be saved because the batch does not contain any profile data."
            << std::endl;
        return;
    }
    cv::imwrite(path, cv::Mat(lineCount, width, CV_8UC1, batch.getIntensityImage().data()));
}
} // namespace

bool acquireProfileDataWithoutCallback(mmind::eye::VirtualProfiler& profiler)
{
    mmind::eye::VirtualUserSet currentUserSet = profiler.currentUserSet();
    int dataPoints = 0;
    // Get the number of data points in each profile
    mmind::eye::showError(currentUserSet.getIntValue(
        mmind::eye::scan_settings::DataPointsPerProfile::name, dataPoints));
    int captureLineCount = 0;
    // Get the value of the "Scan Line Count" parameter
    currentUserSet.getIntValue(mmind::eye::scan_settings::ScanLineCount::name, captureLineCount);

    // Define a ProfileBatch object to store the profile data
    mmind::eye::ProfileBatch profileBatch(dataPoints);

    // Acquire data without using callback
    auto status = profiler.startAcquisition();
    if (!status.isOK()) {
        mmind::eye::showError(status);
        return false;
    }

    while (profileBatch.height() < captureLineCount) {
        // Retrieve the profile data
        mmind::eye::ProfileBatch batch(dataPoints);
        status = profiler.retrieveBatchData(batch);
        if (status.isOK()) {
            if (!profileBatch.append(batch))
                break;
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        } else {
            mmind::eye::showError(status);
            break;
        }
    }

    mmind::eye::showError(profiler.stopAcquisition());
    std::cout << "Save the depth map and intensity image." << std::endl;
    saveMap(profileBatch, profileBatch.height(), profileBatch.width(), "DepthMap.tiff");
    saveIntensity(profileBatch, profileBatch.height(), profileBatch.width(), "IntensityImage.png");
    return true;
}

bool acquireProfileDataWithCallback(mmind::eye::VirtualProfiler& profiler)
{
    mmind::eye::VirtualUserSet currentUserSet = profiler.currentUserSet();
    int dataPoints = 0;
    // Get the number of data points in each profile
    mmind::eye::showError(currentUserSet.getIntValue(
        mmind::eye::scan_settings::DataPointsPerProfile::name, dataPoints));
    int captureLineCount = 0;
    // Get the value of the "Scan Line Count" parameter
    currentUserSet.getIntValue(mmind::eye::scan_settings::ScanLineCount::name, captureLineCount);

    // Define a ProfileBatch object to store the profile data
    mmind::eye::ProfileBatch profileBatch(dataPoints);

    // Acquire data with the callback function
    auto status = profiler.registerAcquisitionCallback(callbackFunc, &profileBatch);
    if (!status.isOK()) {
        mmind::eye::showError(status);
        return false;
    }

    // Call startAcquisition() to enter the virtual device into the acquisition ready status
    status = profiler.startAcquisition();
    if (!status.isOK()) {
        mmind::eye::showError(status);
        return false;
    }

    while (true) {
        std::unique_lock<std::mutex> lock(kMutex);
        if (profileBatch.isEmpty()) {
            lock.unlock();
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        } else
            break;
    }

    mmind::eye::showError(profiler.stopAcquisition());
    std::cout << "Save the depth map and intensity image." << std::endl;
    saveMap(profileBatch, profileBatch.height(), profileBatch.width(),
            "DepthMapUsingCallback.tiff");
    saveIntensity(profileBatch, profileBatch.height(), profileBatch.width(),
                  "IntensityImageUsingCallback.png");
    return true;
}

int main()
{
    try {
        mmind::eye::VirtualProfiler profiler("test.mraw");
        if (!acquireProfileDataWithoutCallback(profiler))
            return -1;
        if (!acquireProfileDataWithCallback(profiler))
            return -1;
    } catch (mmind::eye::ErrorStatus error) {
        mmind::eye::showError(error);
        return -1;
    }
}

File: profiler/use_virtual_device.py
from mecheye.shared import *
from mecheye.profiler import *
from mecheye.profiler_utils import *
import cv2
from time import sleep
from multiprocessing import Lock

mutex = Lock()


class CustomAcquisitionCallback(AcquisitionCallbackBase):
    def __init__(self, width):
        AcquisitionCallbackBase.__init__(self)
        self.total_batch = ProfileBatch(width)

    def run(self, batch):
        mutex.acquire()
        if (not batch.get_error_status().is_ok()):
            print("Error occurred during data acquisition.")
            show_error(batch.get_error_status())
        self.total_batch.append(batch)
        mutex.release()


class UseVirtualDevice(object):
    def __init__(self):
        #  Please ensure that the file name is encoded in UTF-8 format.
        self.profiler = VirtualProfiler("test.mraw")

    def get_parameters(self):
        current_user_set = self.profiler.current_user_set()

        # Get the number of data points in each profile
        error, self.data_width = current_user_set.get_int_value(
            DataPointsPerProfile.name)
        show_error(error)

        # Get the current value of the "Scan Line Count" parameter
        error, self.capture_line_count = current_user_set.get_int_value(
            ScanLineCount.name)
        show_error(error)

    def capture_without_callback(self):
        # Define a ProfileBatch object to store the profile data
        self.total_batch = ProfileBatch(self.data_width)

        # Acquire data without using callback
        status = self.profiler.start_acquisition()
        if not status.is_ok():
            show_error(status)
            return False

        self.total_batch.reserve(self.capture_line_count)
        while self.total_batch.height() < self.capture_line_count:
            batch = ProfileBatch(self.data_width)
            # Retrieve the profile data
            status = self.profiler.retrieve_batch_data(batch)
            if status.is_ok():
                self.total_batch.append(batch)
                sleep(0.1)
            else:
                show_error(status)
                break
        status = self.profiler.stop_acquisition()
        if not status.is_ok():
            show_error(status)
            return False
        return True

    def capture_with_callback(self):
        self.callback = CustomAcquisitionCallback(
            self.data_width)

        # Acquire data with the callback function
        status = self.profiler.register_acquisition_callback(
            self.callback)
        if not status.is_ok():
            show_error(status)
            return False

        # Call startAcquisition() to enter the virtual device into the acquisition ready status
        status = self.profiler.start_acquisition()
        if not status.is_ok():
            show_error(status)
            return False

        while True:
            mutex.acquire()
            if self.callback.total_batch.is_empty():
                mutex.release()
                sleep(0.1)
            else:
                mutex.release()
                break

        status = self.profiler.stop_acquisition()
        if not status.is_ok():
            show_error(status)
            return False
        return True

    def save_images(self, depth_file_name, intensity_file_name):
        cv2.imwrite(depth_file_name,
                    self.total_batch.get_depth_map().data())
        cv2.imwrite(intensity_file_name,
                    self.total_batch.get_intensity_image().data())

    def save_callback_images(self, depth_file_name, intensity_file_name):
        cv2.imwrite(depth_file_name,
                    self.callback.total_batch.get_depth_map().data())
        cv2.imwrite(intensity_file_name,
                    self.callback.total_batch.get_intensity_image().data())

    def main(self):
        self.get_parameters()
        if not self.capture_without_callback():
            return -1
        print("Save the depth map and the intensity image.")
        self.save_images("DepthMap.tiff", "IntensityImage.png")
        # if not self.capture_with_callback():
        #     return -1
        # print("Save the depth map and the intensity image.")
        # self.save_callback_images("DepthMapUsingCallback.tiff",
        #                           "IntensityImageUsingCallback.png")


if __name__ == '__main__':
    try:
        a = UseVirtualDevice()
        a.main()
    except IOError as error:
        print(error)
